ext/src/ruby_api/wasi_config.rs

```diff
@@ -0,0 +1,283 @@
+use super::root;
+use crate::{error, helpers::WrappedStruct};
+use magnus::{
+    function, gc, method, DataTypeFunctions, Error, Module, Object, RArray, RHash, RString,
+    TypedData,
+};
+use std::cell::RefCell;
+use std::{fs::File, path::PathBuf};
+use wasi_common::pipe::ReadPipe;
+use wasmtime_wasi::{WasiCtx, WasiCtxBuilder};
+
+enum ReadStream {
+    Inherit,
+    Path(RString),
+    String(RString),
+}
+
+impl ReadStream {
+    pub fn mark(&self) {
+        match self {
+            Self::Inherit => (),
+            Self::Path(s) => gc::mark(*s),
+            Self::String(s) => gc::mark(*s),
+        }
+    }
+}
+
+enum WriteStream {
+    Inherit,
+    Path(RString),
+}
+impl WriteStream {
+    pub fn mark(&self) {
+        match self {
+            Self::Inherit => (),
+            Self::Path(v) => gc::mark(*v),
+        }
+    }
+}
+
+#[derive(Default)]
+struct WasiConfigInner {
+    stdin: Option<ReadStream>,
+    stdout: Option<WriteStream>,
+    stderr: Option<WriteStream>,
+    env: Option<RHash>,
+    args: Option<RArray>,
+}
+
+macro_rules! mark_some {
```

Comment:
This use-case doesn't seem to warrant a macro, especially since it only appears to be used in this module, or I might be wrong and we could reuse this elsewhere? Can this be replaced by an `iterator` + `for_each`?

Reply:
No strong opinions there, just personally I like to avoid them when possible since they introduce complexity, especially for newcomers.

Reply:
> we could reuse this elsewhere

This feels oddly specific, I don't think we'd use it elsewhere. At least not as written.

🤷 initially I had 5 of these so I thought it made it neater. But with only 3, I agree it's not super valuable. 

> Can this be replaced by an `iterator` + `for_each`?

I'm not sure how to go about that in a efficient manner (and not sure it's worth it?). Maybe we can explore that in our next pairing session, until then I'll remove the macro.

Reply:
> I'm not sure how to go about that in a efficient manner (and not sure it's worth it?)

Not worth it IMO, I added it as an option in case you still want to avoid writing all the statements by hand.